packages/next/src/build/webpack/plugins/slow-module-detection-plugin.ts

```diff
@@ -0,0 +1,330 @@
+
+      if (!moduleName) {
+        return formatChildModules(node, depth)
+      }
+
+      const truncatedName = truncatePath(moduleName)
+      const indentation = TreeSymbols.VERTICAL_LINE.repeat(depth)
+
+      const duration =
+        node.buildDuration > 0
+          ? `(${ConsoleColors.BRIGHT_YELLOW}${formatDuration(node.buildDuration)}${ConsoleColors.RESET})`
+          : ''
+      const nameWithStyle = `${ConsoleColors.BOLD_BLUE}${truncatedName}${ConsoleColors.RESET}`
+      const line = `${indentation}${TreeSymbols.BRANCH}${nameWithStyle} ${duration}`
+
+      return ` ${line}\n${formatChildModules(node, depth + 1)}`
+    }
+
+    // Format all children of a module
+    const formatChildModules = (node: ModuleNode, depth: number): string => {
+      const children = this.moduleChildren.get(node)
+      if (!children) return ''
+
+      return [...children]
+        .map(([_, child]) => formatModuleNode(child, depth))
+        .join('')
+    }
+
+    // Build the complete report
+    let report = ''
+    for (const root of rootModules) {
+      report += formatModuleNode(root, 0)
+    }
+
+    // Only output if there are slow modules to report
+    if (report) {
+      console.log(
+        `${ConsoleColors.BOLD_GREEN}🐌 Detected slow modules while compiling ${this.options.compilerType}:${ConsoleColors.RESET}`
+      )
+      console.log(report)
+    }
+  }
+}
+
+/**
+ * Webpack plugin that analyzes and reports modules that are slow to build.
+ * Hooks into webpack's compilation process to:
+ * 1. Track when module builds start
+ * 2. Record build durations when modules complete
+ * 3. Generate a report when compilation is done
+ */
+export default class SlowModuleDetectionPlugin {
+  private analyzer: ModuleBuildTimeAnalyzer | null = null
+  private moduleBuildStartTimes = new WeakMap<Module, number>()
+
+  constructor(private options: ModuleBuildTimeAnalyzerOptions) {}
+
+  apply = (compiler: Compiler) => {
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+      this.analyzer = new ModuleBuildTimeAnalyzer(this.options)
+
+      // Record start time when module build begins
+      compilation.hooks.buildModule.tap(PLUGIN_NAME, (module) => {
+        this.moduleBuildStartTimes.set(module, Date.now())
+      })
+
+      // Calculate and record duration when module build succeeds
+      compilation.hooks.succeedModule.tap(PLUGIN_NAME, (module) => {
+        const startTime = this.moduleBuildStartTimes.get(module)
+        if (!startTime) {
+          throw new Error(
+            '[SlowModuleDetectionPlugin] Module build timing error: Unable to find the start time for a module build.'
+          )
+        }
+        if (!this.analyzer) {
+          throw new Error(
+            '[SlowModuleDetectionPlugin] Module analyzer initialization error: The analyzer was not properly initialized.'
+          )
+        }
+        this.analyzer.recordModuleBuildTime(
+          module,
+          Date.now() - startTime,
```

Comment:
All performance related time measurements should use `performance.now()`. This is a good habit to get into since `performance.now()` uses a monotonic clock.